### DISCUSSION TITLE
[Core Protocol] Why have the L2 Smart Contracts been implemented as predeploys instead of precompiles?

### DISCUSSION THREAD

**Discussion by PaulRBerg** [ethereum-optimism/developers#418] [Core Protocol] [answered]
### Did you check the documentation?

- [X] Yes

### Did you read the specs?

- [X] Yes
- [ ] No

### Did you check for duplicate questions?

- [X] Yes

### Issue Description

I was looking at the [Smart Contracts overview](https://docs.optimism.io/stack/protocol/rollup/smart-contracts) in the docs, and I was wondering - why have the L2 contracts been implemented as _predeploys_ instead of _precompiles_?

That is, why write these in Solidity instead of in the client software (in `op-geth`)? Wouldn't it have been more efficient and less error-prone to do that? I imagine that upgrading the client software is easier than upgrading smart contracts via the upgradeability pattern.

- BaseFeeVault
- GasPriceOracle
- L1Block
- L1FeeVault
- L2CrossDomainMessenger
- L2ERC721Bridge
- L2StandardBridge
- L2ToL1MessagePasser
- SequencerFeeVault ​

**Comment by hash1go** [comment]
Hello.
Perhaps the decision was made to have the same specifications as other contracts for better interoperability.
However, I would need to ask the engineers on the OP team to know for sure. If resolving this question is a high priority, I will ask the OP team.

**Reply by PaulRBerg** [reply]
Got it.

Not high-priority, no. But thanks for offering to ask.

**Comment by PaulRBerg** [comment]
Could the rationale behind this have been to make Optimism easier to fault proof via [Cannon](https://github.com/ethereum-optimism/cannon)?

cc @IaroslavMazur

**Comment by sbvegan** [comment, marked as answer]
I suspect part of the reason is to not modify the EVM further. A guiding development foundation is [evm equivalence](https://specs.optimism.io/background.html#what-is-evm-equivalence) and handling the L2 smart contracts outside of the evm adheres to that.

**Reply by PaulRBerg** [reply]
Thank you for your answer.

**Reply by sbvegan** [reply]
Yeah sorry for the late reply. I did confirm this is the reason with an engineer internally.
